crates/accelerate/src/twirling.rs

```diff
@@ -0,0 +1,689 @@
+            StandardGate::ZGate,
+        ],
+        0.,
+    ),
+    (
+        [
+            StandardGate::XGate,
+            StandardGate::YGate,
+            StandardGate::YGate,
+            StandardGate::XGate,
+        ],
+        0.,
+    ),
+    (
+        [
+            StandardGate::XGate,
+            StandardGate::IGate,
+            StandardGate::ZGate,
+            StandardGate::YGate,
+        ],
+        0.,
+    ),
+    (
+        [
+            StandardGate::XGate,
+            StandardGate::ZGate,
+            StandardGate::IGate,
+            StandardGate::YGate,
+        ],
+        0.,
+    ),
+    (
+        [
+            StandardGate::XGate,
+            StandardGate::XGate,
+            StandardGate::XGate,
+            StandardGate::XGate,
+        ],
+        0.,
+    ),
+    (
+        [
+            StandardGate::YGate,
+            StandardGate::IGate,
+            StandardGate::ZGate,
+            StandardGate::XGate,
+        ],
+        PI,
+    ),
+    (
+        [
+            StandardGate::YGate,
+            StandardGate::ZGate,
+            StandardGate::IGate,
+            StandardGate::XGate,
+        ],
+        PI,
+    ),
+    (
+        [
+            StandardGate::YGate,
+            StandardGate::XGate,
+            StandardGate::XGate,
+            StandardGate::YGate,
+        ],
+        0.,
+    ),
+    (
+        [
+            StandardGate::YGate,
+            StandardGate::YGate,
+            StandardGate::YGate,
+            StandardGate::YGate,
+        ],
+        0.,
+    ),
+];
+
+#[pyfunction]
+#[pyo3(signature=(circ, twirled_gate, seed=None, num_twirls=1))]
+pub fn twirl_circuit(
+    py: Python,
+    circ: &CircuitData,
+    twirled_gate: StandardGate,
+    seed: Option<u64>,
+    num_twirls: usize
+) -> PyResult<Vec<CircuitData>> {
+    let mut rng = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed),
+        None => Pcg64Mcg::from_entropy(),
+    };
+    let twirl_set: &[([StandardGate; 4], f64)] = match twirled_gate {
+        StandardGate::CXGate => &CX_TWIRL_SET,
+        StandardGate::CZGate => &CZ_TWIRL_SET,
+        StandardGate::ECRGate => &ECR_TWIRL_SET,
+        StandardGate::ISwapGate => &ISWAP_TWIRL_SET,
+        _ => {
+            return Err(QiskitError::new_err(
+                "Provided gate to twirl is not currently supported you can only use CX or CZ",
+            ))
+        }
+    };
+    let generate_twirled_circuit = |rng: &mut Pcg64Mcg| {
+        let mut out_circ = CircuitData::clone_empty_from(circ);
+
+        for inst in circ.data() {
+            match inst.op.view() {
+                OperationRef::Standard(gate) => {
+                    if gate == twirled_gate {
+                        let qubits: Vec<Qubit> = out_circ.get_qargs(inst.qubits).to_vec();
+                        let (twirl, twirl_phase) = twirl_set.choose(rng).unwrap();
+                        out_circ.push_standard_gate(twirl[0], &[], &[qubits[0]])?;
```

Comment:
Would there be a performance benefit to omitting identity gates?

Reply:
I think it wouldn't make much of a difference for the performance of this function. The two ways to do it would be to condition the `push` call on the value of `twirl[index]` which adds a bunch of branching logic around each push call or somehow tweak the static twirling sets which would be tricky because they work well because everything is fixed sizes. There is a potential benefit downstream of this function because we've potentially used less gates

Reply:
Those kinds of nice things have a very poor effect on performance.  We've just made a big deal about Qiskit performance, and if we want to maintain it, we need to keep putting performance first, and designing our interfaces around that.

We can potentially allow it to be opt-in to show the twirling gates, but it's fairly trivial to write the twirling algorithm yourself if that's what you care about seeing, and that's always going to give you the most control over the output.  If what we're after is that kind of "utility" function rather than a high-performance one, I'm not sure it fits in Qiskit - we already shipped off bunches of `qiskit_algorithms` and the applications libraries out of the core library because they didn't fit the "highest performance core data structures and compilation" model, and decided to do that with documentation instead.  If the purpose of the function is educational, then it feels to me like it fits in that vein, and should be handled with guide-level documentation on "introduction to Pauli twirling" rather than being presented in the core library as "this is an efficient way to do it".

Reply:
Sure - I commented on that in my main review.  It's not really clear to me what the purpose of this function is at all from a performance perspective.

If the purpose is to run it post-compilation, then we should design the interface of the function to be smarter about twirls in the first place, so it outputs at least hardware-compatible circuits.  Technically there'll still be possibilities for post-optimisation, since 1q runs of gates will likely be possible to be resynthesised, which again is something we might want to consider.